central/metrics/custom/singleton.go

```diff
@@ -1,11 +1,102 @@
-package aggregator
+package custom
 
 import (
+	"context"
+	"net/http"
+
+	configDS "github.com/stackrox/rox/central/config/datastore"
+	deploymentDS "github.com/stackrox/rox/central/deployment/datastore"
+	"github.com/stackrox/rox/central/metrics"
 	"github.com/stackrox/rox/central/metrics/custom/image_vulnerabilities"
+	custom "github.com/stackrox/rox/central/metrics/custom/tracker"
 	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/grpc/authn"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sync"
+	"github.com/travelaudience/go-promhttp"
+)
+
+var (
+	runner    *aggregatorRunner
+	oneRunner sync.Once
+
+	log = logging.LoggerForModule()
 )
 
-func ValidateConfiguration(config *storage.PrometheusMetrics) error {
-	return image_vulnerabilities.ValidateConfiguration(
-		config.GetImageVulnerabilities().GetDescriptors())
+type aggregatorRunner struct {
+	http.Handler
+	image_vulnerabilities custom.Tracker
+}
+
+type Runner interface {
+	http.Handler
+	ValidateConfiguration(*storage.PrometheusMetrics) (*RunnerConfiguration, error)
+	Reconfigure(*RunnerConfiguration)
+}
+
+// Singleton returns a runner, or nil if there were errors durig initialization.
+// nil runner is safe, but no-op.
+func Singleton() Runner {
+	oneRunner.Do(func() {
+		runner = makeRunner(metrics.MakeCustomRegistry(), deploymentDS.Singleton())
+		go runner.initialize(configDS.Singleton())
+	})
+	return runner
+}
+
+func makeRunner(registry metrics.CustomRegistry, dds deploymentDS.DataStore) *aggregatorRunner {
+	return &aggregatorRunner{
+		Handler:               promhttp.HandlerFor(registry, promhttp.HandlerOpts{}),
+		image_vulnerabilities: image_vulnerabilities.New(registry, dds),
+	}
+}
+
+func (ar *aggregatorRunner) initialize(cds configDS.DataStore) {
+	ctx := sac.WithAllAccess(context.Background())
+	systemPrivateConfig, err := cds.GetPrivateConfig(ctx)
+	if err != nil {
+		log.Errorw("Failed to read Prometheus metrics configuration from the DB", logging.Err(err))
+		return
+	}
+
+	cfg, err := ar.ValidateConfiguration(systemPrivateConfig.GetMetrics())
+	if err != nil {
+		log.Errorw("Failed to configure Prometheus metrics", logging.Err(err))
+		return
+	}
+	ar.Reconfigure(cfg)
+}
+
+// RunnerConfiguration is to pass between ParseConfiguration and Reconfigure.
```

Comment:
Why did you include `RunnerConfiguration` in `singleton.go`? I would have extracted that to a different file to keep the singleton file minimal.